examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+"""
+Title: Automatic Speech Recognition with Transformer
+Author: [Apoorv Nandan](https://twitter.com/NandanApoorv)
+Date created: 2021/01/13
+Last modified: 2021/01/13
+Description: Implement a complete transformer and train it on sequence to sequence ASR.
+"""
+"""
+## Introduction
+
+Automatic speech recognition (ASR) consists of transcribing audio segments into text.
+A popular way to handle this task is to treat it as a sequence-to-sequence
+problem. The audio can be represented as a sequence of feature vectors,
+and the text can be represented as a sequence of characters, words, or subword tokens.
+
+For this demonstration, we will use LJSpeech dataset from the
+[LibriVox](https://librivox.org/) project. It consists of short
+audio clips of a single speaker reading passages from 7 non-fiction books.
+Our model will be similar to the original Transformer (both encoder and decoder)
+as proposed in the paper, "Attention is All You Need".
+
+
+**References:**
+- [Attention is All You Need](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)
+- [Very Deep Self-Attention Networks for End-to-End Speech Recognition](https://arxiv.org/pdf/1904.13377.pdf)
+- [Speech Transformers](https://ieeexplore.ieee.org/document/8462506)
+- [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
+"""
+
+
+import os
+import random
+from glob import glob
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+
+"""
+## Transformer Input Layer
+
+When processing past target tokens for the decoder, we compute the sum of 
+position embeddings and token embeddings.
+
+When processing audio features, we apply convolutional layers to downsample
+them and process local relationships. It makes the training stable.
+"""
+
+
+class TransformerInput(layers.Layer):
```

Comment:
The use of `Input` can be confusing due to the collision with `keras.Input`. I would recommend calling this `PositionalEmbedding` or something similar, as well as separating the code for the 2 input types, since in practice they have no overlap. This class will be more readable if it only covers the "tokens" mode.

The "features" mode probably doesn't require its own class, you could just inline the conv1d layers in the `Transformer` class.